Userland/Libraries/LibJS/Console.cpp

```diff
@@ -6,14 +6,110 @@
+            return js_undefined();
+        }
+
+        if (log_level == Console::LogLevel::Group || log_level == Console::LogLevel::GroupCollapsed) {
+            auto& group = arguments.get<Console::Group>();
+            m_queued_commands.empend(ConsoleCommandType::Group, log_level, group.label);
+            return js_undefined();
+        }
+
+        auto& logging_arguments = arguments.get<Vector<Value>>();
+        MarkedVector<Value> marked_arguments { vm().heap() };
+        marked_arguments.extend(logging_arguments);
+
+        m_queued_commands.empend(ConsoleCommandType::Printer, log_level, String {}, move(marked_arguments));
+        return js_undefined();
+    }
+
+    virtual void clear() override
+    {
+        m_queued_commands.empend(ConsoleCommandType::Clear);
+    }
+    virtual void end_group() override
+    {
+        m_queued_commands.empend(ConsoleCommandType::EndGroup);
+    }
+
+    void run_commands(ConsoleClient& client)
+    {
+
+        if (!m_queued_commands.is_empty())
+            dbgln("\033[32;1m(Outputting queued commands from js log)\033[0m");
+
+        // We just ignore the results here as we have already return undefined.
+        for (auto& command : m_queued_commands) {
+            switch (command.type) {
+            case ConsoleCommandType::Trace:
+                VERIFY(command.arguments.has<Vector<String>>());
+                MUST(client.printer(command.log_level, Console::Trace { move(command.string_data), move(command.arguments.get<Vector<String>>()) }));
+                break;
+            case ConsoleCommandType::Group:
+                MUST(client.printer(command.log_level, Console::Group { move(command.string_data) }));
+                break;
+            case ConsoleCommandType::Printer: {
+                VERIFY(command.arguments.has<MarkedVector<Value>>());
+                auto& value_arguments = command.arguments.get<MarkedVector<Value>>();
+                Vector<Value> raw_values;
+                raw_values.extend(value_arguments);
+                MUST(client.printer(command.log_level, raw_values));
+                break;
+            }
+            case ConsoleCommandType::Clear:
+                client.clear();
+                break;
+            case ConsoleCommandType::EndGroup:
+                client.end_group();
+                break;
+            }
+        }
+        m_queued_commands.clear();
+    }
+
+private:
+    enum class ConsoleCommandType {
+        Trace,
+        Group,
+        Printer,
+        Clear,
+        EndGroup
+    };
+
+    struct ConsoleCommand {
+        ConsoleCommandType type;
+        Console::LogLevel log_level = Console::LogLevel::Assert;
+        String string_data {};
+        Variant<Empty, Vector<String>, MarkedVector<Value>> arguments {};
+    };
+
+    Vector<ConsoleCommand> m_queued_commands;
+};
+
 Console::Console(GlobalObject& global_object)
     : m_global_object(global_object)
+    , m_client(new QueueingConsoleClient(*this))
```

Comment:
This raw `new` (and the `delete` below) irk me a great deal. But unless we want to make ConsoleClient RefCounted I don't know how to nicely solve this.

(I don't know why it is a raw pointer currently but didn't just want to change that)

Reply:
And this immediately caused problems with leaks, maybe we should make it a RefPtr after all...